src/async/imap/MCIMAPAsyncSession.cpp

```diff
@@ -306,58 +315,55 @@ IMAPAsyncConnection * IMAPAsyncSession::sessionForFolder(String * folder, bool u
 {
-    if (mMaximumConnections == 0) {
-        for(unsigned int i = 0 ; i < mSessions->count() ; i ++) {
-            IMAPAsyncConnection * s = (IMAPAsyncConnection *) mSessions->objectAtIndex(i);
-            if (s->operationsCount() == 0)
-                return s;
-        }
-        IMAPAsyncConnection * chosenSession = session();
-        mSessions->addObject(chosenSession);
+    IMAPAsyncConnection * chosenSession = sessionWithMinQueue();
+    if ((chosenSession != NULL) && (chosenSession->operationsCount() == 0)) {
         return chosenSession;
     }
-    else {
-        IMAPAsyncConnection * chosenSession = NULL;
-        unsigned int minOperationsCount = 0;
-        for(unsigned int i = 0 ; i < mSessions->count() ; i ++) {
-            IMAPAsyncConnection * s = (IMAPAsyncConnection *) mSessions->objectAtIndex(i);
-            if (chosenSession == NULL) {
-                chosenSession = s;
-                minOperationsCount = s->operationsCount();
-            }
-            else if (s->operationsCount() < minOperationsCount) {
-                chosenSession = s;
-                minOperationsCount = s->operationsCount();
-            }
-        }
-        if (mSessions->count() < mMaximumConnections) {
-            if ((chosenSession != NULL) && (minOperationsCount == 0)) {
-                return chosenSession;
-            }
-            chosenSession = session();
-            mSessions->addObject(chosenSession);
-            return chosenSession;
-        }
-        else {
-            return chosenSession;
-        }
+
+    if ((mMaximumConnections == 0) || (mSessions->count() < mMaximumConnections)) {
+        chosenSession = session();
+        mSessions->addObject(chosenSession);
     }
+
+    return chosenSession;
 }
 
 IMAPAsyncConnection * IMAPAsyncSession::matchingSessionForFolder(String * folder)
 {
-    for(unsigned int i = 0 ; i < mSessions->count() ; i ++) {
-        IMAPAsyncConnection * currentSession = (IMAPAsyncConnection *) mSessions->objectAtIndex(i);
-        if (currentSession->lastFolder() != NULL) {
-            if (currentSession->lastFolder()->isEqual(folder)) {
-                return currentSession;
+    IMAPAsyncConnection * chosenSelectedSession = sessionWithMinQueue(true, folder);
+    if (chosenSelectedSession != NULL) {
+        return chosenSelectedSession;
+    }
+
+    IMAPAsyncConnection * chosenNonSelectedSession = sessionWithMinQueue(true, NULL);
+    if (chosenNonSelectedSession != NULL) {
+        return chosenNonSelectedSession;
+    }
+
+    return availableSession();
+}
+
+IMAPAsyncConnection * IMAPAsyncSession::sessionWithMinQueue(bool filterByFolder, String * folder)
+{
+    IMAPAsyncConnection * chosenSession = NULL;
+    unsigned int minOperationsCount = UINT_MAX;
+
+    for (unsigned int i = 0 ; i < mSessions->count() ; i ++) {
+        IMAPAsyncConnection * s = (IMAPAsyncConnection *) mSessions->objectAtIndex(i);
+        if (s->operationsCount() < minOperationsCount) {
```

Comment:
could you change it to if ((chosenSession == NULL) || (s->operationsCount() < minOperationsCount) )
then, initialize minOperationsCount to zero.

and remove `#include <limits.h>`